README
mod_caldav:
-----------
-this is an Apache module implementation of RFC 4791, i.e. adding CalDAV
support to WebDAV
-depends on mod_dav_acl <http://sourceforge.net/projects/moddavacl/>,
which implements RFC 3744 also as an Apache module
-uses libical for iCalendar manipulations
<http://sourceforge.net/projects/freeassociation/>
Before release 0.1.8 the fork of libical in Evolution Data Server (EDS) was
used. (this module does not use other components of it)
-included some patches to EDS: e.g. some heap bug fixes in libical and also
client side fixes for using Evolution as a calendar client: caches timezone
info for scheduled events, forward proxy enabled, etc. The latest svn version
of eds seems to be quite reasonable now, however. There are no patches to libical
of <http://sourceforge.net/projects/freeassociation/> version (since it already
has these fixes).

Configuration:
--------------
-must be loaded (dlopen) after mod_dav_acl and mod_dav
-once the module installs, some directives from mod_dav_acl are needed:

DAVETagResponse on
-enable strong ETag responses

DAVACL on
-enable ACL handling

and e.g. setting (although it is not used anyway by the current implementation)
MaxAttendeesPerInstance   10000

in debian:"sudo a2enmod dav_caldav" enables module loading and basic
configuration

Install:
--------
-the module is debianized so dpkg-buildpackage should work at least in ubuntu
-otherwise "standard" ./configure && make && make install. You may need to run
autogen.sh to generate a proper ./configure etc.

Tests:
------
-once the module is successfully compiled and loaded, tests can be run with
make check or by running ./caldav_tests. You need to have admin (digest)
account and /xxx/caldav directory must be available and accessible by apache

Performance:
------------
-enabling acl handling certainly decreases throughput. Especially after
each PUT, the owner of a resource is stored. Current mod_dav uses db4 database
for property storings and it seems to be quite slow, so PUT performance may
degrade a lot. You can however, enable owner storage onto extended attributes
which is much faster.


ChangeLog
23.09.2008 Jari Urpalainen
	* caldav_ical.[ch], configure.in:
	-changed to use the mainstream libical from <http://sourceforge.net/projects/freeassociation/>
	-removed eds-libical hacks

08.07.2008 Jari Urpalainen
	* caldav.c:
	-added conditional report responses based on a collection ETag:
	If-None-Match: "abcd2333" (with the same ETag) is reponded by 304 Not Modified

06.06.2008 Jari Urpalainen
	* caldav_ical.[ch]:
	-added hacks for eds-libical abi changes
	-fixed wrong search order in strstr

19.02.2008 Jari Urpalainen
	* caldav.c:
	-fixed Depth header handling
	-fixed body reading with multiple threads
	
07.08.2007 Jari Urpalainen
	* caldav.c:
	-fixed a bug when storing resourcetype property with mkcalendar method

04.07.2007 Jari Urpalainen
	* configure.in:
	-libical from evolution-data-server may be replaced with another fork.
	(DEPS_LIBICAL_LIBS=-lical DEPS_LIBICAL_CFLAGS=I/xxxx in configure)
	(caveat: some tests may fail, however ...)	
	* test-cli.c:
	-added libcurl based small http client for testing purposes
	* caldav.c:
	-fixed a memory leak in filtering function

29.03.2007 Jari Urpalainen
        * caldav_files:
        -added more tests from rfc4791
	* caldav_ical.c:
	-fixed due evaluation in vtodos
	* caldav.c:
	-fixed multiget with full uris
	* testical.c:
	-added for testing simple report requests over an iCalendar file

28.03.2007 Jari Urpalainen
	* caldav_ical.c:
	-memory bug in expanded recurrence set fixed
	-recurring freebusy report fix 	
	
27.03.2007 Jari Urpalainen	
	* caldav_ical.c:
	-in free_busy_query report return utc times
	* caldav.c:
	-in free_busy_query report return also mandatory dtstamp

26.03.2007 Jari Urpalainen
	* Initial release 0.1.0
	


caldav.h
/**
 * This is part of a mod_caldav library.
 *
 * Copyright (C) 2006-2009 Nokia Corporation.
 *
 * Contact: Jari Urpalainen <jari.urpalainen@nokia.com>
 *
 * This library is free software; you can redistribute it and/or
 * modify it under the terms of the GNU Lesser General Public
 * License as published by the Free Software Foundation; either
 * version 2.1 of the License, or (at your option) any later version.
 *
 * This library is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * Lesser General Public License for more details.

 * You should have received a copy of the GNU Lesser General Public
 * License along with this library; if not, write to the Free Software
 * Foundation, Inc., 51 Franklin St, Fifth Floor, Boston, MA  02110-1301  USA
 *
 */

#ifndef _CALDAV_H_
#define _CALDAV_H_

const char *caldav_ctag(request_rec *r);

#endif /* _CALDAV_H_ */

mod_caldav.h
/**
 * This is part of a mod_caldav library.
 *
 * Copyright (C) 2006 Nokia Corporation.
 *
 * Contact: Jari Urpalainen <jari.urpalainen@nokia.com>
 *
 * This library is free software; you can redistribute it and/or
 * modify it under the terms of the GNU Lesser General Public
 * License as published by the Free Software Foundation; either
 * version 2.1 of the License, or (at your option) any later version.
 *
 * This library is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * Lesser General Public License for more details.

 * You should have received a copy of the GNU Lesser General Public
 * License along with this library; if not, write to the Free Software
 * Foundation, Inc., 51 Franklin St, Fifth Floor, Boston, MA  02110-1301  USA
 *
 */
#ifndef _MOD_CALDAV_H_
#define _MOD_CALDAV_H_

#define NS_CALDAV "urn:ietf:params:xml:ns:caldav"

#define XML_VERSION "1.0"

#define NODE_NS(node, ns_string) (node->ns && node->ns->href && \
				  strcmp((char *) node->ns->href, ns_string) == 0)

#define FOR_CHILD(node, parent) \
	for (node = parent ? parent->children : NULL; node; node = node->next)

#define NODE_NOT_DAV(node) node->type != XML_ELEMENT_NODE || !NODE_NS(node, NS_DAV)
#define NODE_NOT_CALDAV(node) node->type != XML_ELEMENT_NODE || !NODE_NS(node, NS_CALDAV)

#define NODE_MATCH(node, y) (strcmp((char *) node->name, y) == 0)

typedef struct caldav_search_s caldav_search_t;


#endif /* _MOD_CALDAV_H_ */


caldav_ical.h
/**
 * This is part of a mod_caldav library.
 * 
 * Copyright (C) 2006 Nokia Corporation.
 * 
 * Contact: Jari Urpalainen <jari.urpalainen@nokia.com>
 * 
 * This library is free software; you can redistribute it and/or
 * modify it under the terms of the GNU Lesser General Public
 * License as published by the Free Software Foundation; either
 * version 2.1 of the License, or (at your option) any later version.
 * 
 * This library is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * Lesser General Public License for more details.

 * You should have received a copy of the GNU Lesser General Public
 * License along with this library; if not, write to the Free Software
 * Foundation, Inc., 51 Franklin St, Fifth Floor, Boston, MA  02110-1301  USA
 * 
 */

#ifndef _CALDAV_ICAL_H_
#define _CALDAV_ICAL_H_

void caldav_ical_freebusy(const char *file,
			  const icaltimezone *tz,
			  icaltimetype start,
			  icaltimetype end,
			  icalcomponent *freebusy);

int caldav_ical_timerange(xmlNodePtr child,
			  icaltimetype *pstart,
			  icaltimetype *pend);

int caldav_ical_search(const char *file,
			const icaltimezone *tz,
			xmlNodePtr caldata,
			xmlNodePtr filter,
			caldav_search_t **p);

char *caldav_ical_dump(caldav_search_t *p);
int caldav_ical_free(caldav_search_t *p);

#endif  /* _CALDAV_ICAL_H_ */


caldav_liveprops.c
/**
 * This is part of a mod_caldav library.
 *
 * Copyright (C) 2006 Nokia Corporation.
 *
 * Contact: Jari Urpalainen <jari.urpalainen@nokia.com>
 *
 * This library is free software; you can redistribute it and/or
 * modify it under the terms of the GNU Lesser General Public
 * License as published by the Free Software Foundation; either
 * version 2.1 of the License, or (at your option) any later version.
 *
 * This library is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * Lesser General Public License for more details.

 * You should have received a copy of the GNU Lesser General Public
 * License along with this library; if not, write to the Free Software
 * Foundation, Inc., 51 Franklin St, Fifth Floor, Boston, MA  02110-1301  USA
 *
 */
/*
 * caldav_liveprops.c: mod_caldav live property provider functions
 *
 */

#include <sys/types.h>

#include <httpd.h>
#include <libxml/tree.h>

#include <mod_dav.h>
#include "caldav.h"

#include "mod_dav_acl.h"

#include <libical/ical.h>
#include "mod_caldav.h"
#include "apr_strings.h"
#include "caldav_ical.h"

#define NS_CTAG "http://calendarserver.org/ns/"

/*
** The namespace URIs that we use. This list and the enumeration must
** stay in sync.
*/
static const char * const caldav_namespace_uris[] =
{
    NS_CALDAV,
    NS_CTAG,
    NULL        /* sentinel */
};

enum {
    CALDAV_NAMESPACE_URI_NO = 0,  /* the CALDAV: namespace URI ref number */
    CALDAV_NAMESPACE_CTAG_NO      /* the CTAG: namespace URI ref number */
};

#define CALDAV_RO_PROP(name, enum_name) \
        { CALDAV_NAMESPACE_URI_NO, name, CALDAV_PROPID_##enum_name, 0 }
#define CALDAV_RW_PROP(name, enum_name) \
        { CALDAV_NAMESPACE_URI_NO, name, CALDAV_PROPID_##enum_name, 1 }

enum {
    CALDAV_PROPID_calendar_description = 1,
    CALDAV_PROPID_calendar_timezone,
    CALDAV_PROPID_supported_calendar_component_set,
    CALDAV_PROPID_supported_calendar_data,
    CALDAV_PROPID_max_resource_size,
    CALDAV_PROPID_min_date_time,
    CALDAV_PROPID_max_date_time,
    CALDAV_PROPID_max_instances,
    CALDAV_PROPID_max_attendees_per_instance,
    CALDAV_PROPID_calendar_data,
    CALDAV_PROPID_calendar_home_set,
    CALDAV_PROPID_getctag
};

static const dav_liveprop_spec caldav_props[] =
{
    CALDAV_RO_PROP("supported-calendar-component-set", supported_calendar_component_set),
    CALDAV_RO_PROP("supported-calendar-data", supported_calendar_data),
    CALDAV_RO_PROP("max-resource-size", max_resource_size),
    CALDAV_RO_PROP("min-date-time", min_date_time),
    CALDAV_RO_PROP("max-date-time", max_date_time),
    CALDAV_RO_PROP("max-instances", max_instances),
    CALDAV_RO_PROP("max-attendees-per-instance", max_attendees_per_instance),
    CALDAV_RW_PROP("calendar-description", calendar_description),
    CALDAV_RW_PROP("calendar-timezone", calendar_timezone),
    CALDAV_RO_PROP("calendar-data", calendar_data),
    CALDAV_RW_PROP("calendar-home-set", calendar_home_set),

    { CALDAV_NAMESPACE_CTAG_NO, "getctag", CALDAV_PROPID_getctag, 0 },

    { 0 } /* sentinel */
};

const dav_hooks_liveprop caldav_hooks_liveprop;

static const dav_liveprop_group caldav_liveprop_group =
{
    caldav_props,
    caldav_namespace_uris,
    &caldav_hooks_liveprop
};


static dav_prop_insert caldav_insert_prop(const dav_resource *resource,
						int propid,
						dav_prop_insert what,
						apr_text_header *phdr)
{
    const char *value = NULL;
    const char *s = NULL;
    apr_pool_t *p = resource->pool;
    const dav_liveprop_spec *info;
    int global_ns;
    char *fval = NULL;

    if (!resource->exists)
	return DAV_PROP_INSERT_NOTDEF;

    /* ### we may want to respond to DAV_PROPID_resourcetype for PRIVATE
       ### resources. need to think on "proper" interaction with mod_dav */

    switch (propid) {
    case CALDAV_PROPID_getctag:
	{
	    request_rec *r = resource->hooks->get_request_rec(resource);

	    value = caldav_ctag(r);
	    if (value == NULL)
		return DAV_PROP_INSERT_NOTDEF;
	}
	break;

    case CALDAV_PROPID_calendar_data:
	{
	    request_rec *r = resource->hooks->get_request_rec(resource);
	    caldav_search_t *p;

	    if (r->method_number != M_REPORT)
		return DAV_PROP_INSERT_NOTDEF;

	    p = resource->ctx;
	    value = fval = caldav_ical_dump(p);
	    if (value == NULL)
		return DAV_PROP_INSERT_NOTDEF;
	    break;
	}

    default:
	/* ### what the heck was this property? */
	return DAV_PROP_INSERT_NOTDEF;
    }

    /* assert: value != NULL */

    /* get the information and global NS index for the property */
    global_ns = dav_get_liveprop_info(propid, &caldav_liveprop_group, &info);

    /* assert: info != NULL && info->name != NULL */

    if (what == DAV_PROP_INSERT_VALUE)
	s = apr_psprintf(p, "<lp%d:%s>%s</lp%d:%s>" DEBUG_CR,
			 global_ns, info->name, value, global_ns, info->name);
    else if (what == DAV_PROP_INSERT_NAME)
	s = apr_psprintf(p, "<lp%d:%s/>" DEBUG_CR, global_ns, info->name);
    else
	/* assert: what == DAV_PROP_INSERT_SUPPORTED */
	s = apr_psprintf(p, "<D:supported-live-property D:name=\"%s\" "
			    "D:namespace=\"%s\"/>" DEBUG_CR,
			    info->name, caldav_namespace_uris[info->ns]);

    apr_text_append(p, phdr, s);

    free(fval);
    /* we inserted whatever was asked for */
    return what;
}

static int caldav_is_writable (const dav_resource *resource, int propid)
{
    const dav_liveprop_spec *info;

    dav_get_liveprop_info(propid, &caldav_liveprop_group, &info);

    return info->is_writable;
}

static dav_error *caldav_patch_validate(const dav_resource *resource,
					const apr_xml_elem *elem,
					int operation, void **context,
					int *defer_to_dead)
{
    /* NOTE: this function will not be called unless/until we have
       modifiable (writable) live properties. */
    dav_elem_private *priv = elem->priv;

    switch (priv->propid) {
    case CALDAV_PROPID_calendar_description:
	*defer_to_dead = TRUE;
	break;

    case CALDAV_PROPID_calendar_timezone:
	/* actually a dead property, but defined as alive to have
	 * this callback for proppatch
	 */
	*defer_to_dead = TRUE;
	break;

    case CALDAV_PROPID_calendar_home_set:
	if (!dav_acl_is_resource_principal(resource)) {
	    return dav_new_error(resource->pool, HTTP_CONFLICT, 0,
				 "The resource URI is not a principal");
	}
	*defer_to_dead = TRUE;
	break;

    default:
	break;
    }
    return NULL;
}

static dav_error *caldav_patch_exec(const dav_resource *resource,
					const apr_xml_elem *elem,
					int operation, void *context,
					dav_liveprop_rollback **rollback_ctx)
{
    /* NOTE: this function will not be called unless/until we have
       modifiable (writable) live properties. */
    return NULL;
}

static void caldav_patch_commit(const dav_resource *resource,
				int operation, void *context,
				dav_liveprop_rollback *rollback_ctx)
{
    /* NOTE: this function will not be called unless/until we have
       modifiable (writable) live properties. */
}

static dav_error *caldav_patch_rollback(const dav_resource *resource,
					int operation, void *context,
					dav_liveprop_rollback *rollback_ctx)
{
    /* NOTE: this function will not be called unless/until we have
       modifiable (writable) live properties. */
    return NULL;
}

const dav_hooks_liveprop caldav_hooks_liveprop = {
    caldav_insert_prop,
    caldav_is_writable,
    caldav_namespace_uris,
    caldav_patch_validate,
    caldav_patch_exec,
    caldav_patch_commit,
    caldav_patch_rollback,
};

void caldav_gather_propsets(apr_array_header_t *uris)
{
}

int caldav_find_liveprop(const dav_resource *resource,
				const char *ns_uri, const char *name,
				const dav_hooks_liveprop **hooks)
{
    /* don't try to find any liveprops if this isn't "our" resource
	if (resource->hooks != &caldav_hooks_repos)
	   return 0;
     */
    return dav_do_find_liveprop(ns_uri, name, &caldav_liveprop_group, hooks);
}

void caldav_register_props(apr_pool_t *p)
{
    /* register the namespace URIs */
    dav_register_liveprop_group(p, &caldav_liveprop_group);
}



caldav_liveprops.h
/**
 * This is part of a mod_caldav library.
 * 
 * Copyright (C) 2006 Nokia Corporation.
 * 
 * Contact: Jari Urpalainen <jari.urpalainen@nokia.com>
 * 
 * This library is free software; you can redistribute it and/or
 * modify it under the terms of the GNU Lesser General Public
 * License as published by the Free Software Foundation; either
 * version 2.1 of the License, or (at your option) any later version.
 * 
 * This library is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * Lesser General Public License for more details.

 * You should have received a copy of the GNU Lesser General Public
 * License along with this library; if not, write to the Free Software
 * Foundation, Inc., 51 Franklin St, Fifth Floor, Boston, MA  02110-1301  USA
 * 
 */
#ifndef _CALDAV_LIVEPROPS_H_
#define _CALDAV_LIVEPROPS_H_

void caldav_register_props(apr_pool_t *p);
void caldav_gather_propsets(apr_array_header_t *uris);

int caldav_find_liveprop(const dav_resource *resource,
			 const char *ns_uri, const char *name,
			 const dav_hooks_liveprop **hooks);

#endif  /* _CALDAV_LIVEPROPS_H_ */


testical.c
/**
 * This is part of a mod_caldav library.
 *
 * Copyright (C) 2006 Nokia Corporation.
 *
 * Contact: Jari Urpalainen <jari.urpalainen@nokia.com>
 *
 * This library is free software; you can redistribute it and/or
 * modify it under the terms of the GNU Lesser General Public
 * License as published by the Free Software Foundation; either
 * version 2.1 of the License, or (at your option) any later version.
 *
 * This library is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * Lesser General Public License for more details.

 * You should have received a copy of the GNU Lesser General Public
 * License along with this library; if not, write to the Free Software
 * Foundation, Inc., 51 Franklin St, Fifth Floor, Boston, MA  02110-1301  USA
 *
 */

#include "config.h"

#include <stddef.h>
#include <stdlib.h>
#include <string.h>
#include <stdio.h>
#include <assert.h>
#include <signal.h>
#include <ctype.h>
#include <unistd.h>

#include <libxml/tree.h>

#include <libical/ical.h>

#include "mod_caldav.h"
#include "caldav_ical.h"

#define NS_DAV	"DAV:"

static void usage (const char *argv)
{
    fprintf (stdout, "%s -q request_xml_doc -i icalfile\n" \
		     "returns 0 if succeeds, 1 for an error\n", argv);
}

/** main for libical search testings */
int main (int argc, char *argv[])
{
    int opt, rc;
    const char *file = NULL, *ical = NULL;
    xmlDocPtr doc;
    caldav_search_t *p = NULL;
    xmlNodePtr node, caldata = NULL, child = NULL;

    while ((opt = getopt(argc, argv, "q:i:")) != -1) {
	switch (opt) {
	case ':':
	case 'h':
	case '?':
	    usage(argv[0]);
	    return EXIT_SUCCESS;

	case 'q':
	    file = optarg;
	    break;

	case 'i':
	    ical = optarg;
	    break;
	}
    }
    if (file == NULL || ical == NULL) {
	usage(argv[0]);
	return EXIT_FAILURE;
    }
    /* do not prevent the creation of cdata nodes */
    doc = xmlParseFile(file);

    node = doc ? doc->children : NULL;

    for ( ; node; node = node->next) {
	if (NODE_NOT_CALDAV(node))
	    ;
	else if (NODE_MATCH(node, "calendar-query")) {
	    FOR_CHILD(node, node) {
		if (NODE_NOT_DAV(node))
		    ;
		else if (NODE_MATCH(node, "prop"))
		    break;
	    }
	    break;
	}
    }

    FOR_CHILD(caldata, node) {
	if (NODE_NOT_CALDAV(caldata))
	    ;
	else if (NODE_MATCH(caldata, "calendar-data"))
	    break;
    }
    child = node ? node->parent : NULL;
    FOR_CHILD(child, child) {
	if (NODE_MATCH(child, "filter"))
	    break;
    }

    if ((rc = caldav_ical_search(ical, NULL, caldata,
				 child ? child->children : NULL, &p))) {
	char *pch = NULL;

	printf ("%s\n", pch = caldav_ical_dump(p));
	free(pch);
    }
    else {
	printf("search did not find anything\n");
    }

    caldav_ical_free(p);

    xmlFreeDoc(doc);

    xmlCleanupParser();

    return rc ? EXIT_SUCCESS : EXIT_FAILURE;
}

